### PR TITLE
[Approximation Framework] Enabling approximation on top-level term queries on numeric fields

### LINKED ISSUE #18620: [Approximation Framework] Extend approximation to top-level term queries on numeric fields
### Is your feature request related to a problem? Please describe

Numeric fields like long, int, etc use lucene's `newExactQuery` which return a `newRangeQuery` which is internally a `PointRangeQuery`. With `ApproximatePointRangeQuery` we should be able to wrap top level term queries against these mappers. I think it is fairly common for users today to search for a single term indexed as a number and it should help speed up the results

### Describe the solution you'd like

Wrap existing field mapper term queries for number fields with `ApproximatePointRangeQuery`. `canApproximate` will receive `SearchContext` and will kick in approximation when appropriate

### Related component

Search:Performance

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by sawansri**
### Description
Enables approximation in top level term queries on numeric fields by wrapping `termQuery` mappers in `NumberFieldMapper` with `ApproximatePointRangeQuery` instead of Lucene's `newExactQuery`

### Related Issues
Resolves #18620

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by sawansri** [comment]
I'm waiting on #18530 to get merged before rebasing and refactoring my tests to integrate with the parameterized versions in `ApproximatePointRangeQueryTests`.

**Comment by sawansri** [comment]
Seeing ~25% reduction in latency with approximation

Before approximation
```
|                                        50th percentile latency |   term |     12.1217 |     ms |
|                                        90th percentile latency |   term |     13.4443 |     ms |
|                                        99th percentile latency |   term |     18.9051 |     ms |
|                                       100th percentile latency |   term |     18.9174 |     ms |
|                                   50th percentile service time |   term |     11.1184 |     ms |
|                                   90th percentile service time |   term |      12.592 |     ms |
|                                   99th percentile service time |   term |      16.952 |     ms |
|                                  100th percentile service time |   term |     17.2343 |     ms |
|                                                     error rate |   term |           0 |      % |
```

After approximation
```
|                                        50th percentile latency |   term |     9.45359 |     ms |
|                                        90th percentile latency |   term |     10.2816 |     ms |
|                                        99th percentile latency |   term |     17.5361 |     ms |
|                                       100th percentile latency |   term |     28.0131 |     ms |
|                                   50th percentile service time |   term |     8.63716 |     ms |
|                                   90th percentile service time |   term |      9.1716 |     ms |
|                                   99th percentile service time |   term |     10.7578 |     ms |
|                                  100th percentile service time |   term |     26.8904 |     ms |
|                                                     error rate |   term |           0 |      % |
```

Query:
```
curl -X POST "http://localhost:9200/logs-*/_search" -H "Content-Type: application/json" -d '
{
  "query": {
    "term": {
      "status": {
        "value": 200
      }
    }
  }       
}
' | jq '.'
```

**Comment by atris** [comment]
25% latency reduction is significant. Did you validate this holds across different data distributions? Term queries on sparse vs dense numeric fields might behave differently with approximation.

  `ApproximateTermQuery` wrapping looks clean. Two concerns:
  1. Does this maintain correct scoring semantics for numeric term queries?
  2. How does approximation error propagate in compound queries containing these terms?

  Parameterized tests are good but some edge cases to consider:
  - What happens with numeric overflow scenarios?
  - Are negative numbers handled correctly in the approximation?
  - Need explicit test for approximation vs exact result validation.

  Nice integration with existing approximation framework. Consider:
  - Should this be configurable per query? Some use cases may require exact numeric matching.
  - Memory overhead of maintaining both approximate and exact structures?

  Benchmark methodology is unclear - what dataset size/cardinality? Need to validate performance doesn't regress for small result sets where approximation overhead might dominate.

**Review by atris — CHANGES_REQUESTED** [review]
Please refer to: https://github.com/opensearch-project/OpenSearch/pull/18679#issuecomment-3145315596

**Comment by sawansri** [comment]
> 25% latency reduction is significant. Did you validate this holds across different data distributions? Term queries on sparse vs dense numeric fields might behave differently with approximation.

I agree, I do need to do more testing on higher cardinality fields as well as more balanced datasets to see performance impact. 

> `ApproximateTermQuery` wrapping looks clean. Two concerns:
> 
> 1. Does this maintain correct scoring semantics for numeric term queries?
> 2. How does approximation error propagate in compound queries containing these terms?

So I'm not actually implementing `ApproximateTermQuery` (it would introduce unnecessary overhead). Instead, I'm simply replacing Lucene's `newExactQuery` (internally a `PointRangeQuery` on the same upper and lower field) with an equivalent `ApproximatePointRangeQuery` in `NumberFieldMapper`. 

1. Scoring should not be impacted for term queries on numeric fields since they are constant score queries. Also,  a `PointRangeQuery` and `ApproximatePointRangeQuery` are identical scoring wise.

2. Approximation is not currently applied for any clause in a boolean query or any case where it isn't a top level query. Result inaccuracy should not be an issue here. 

> Parameterized tests are good but some edge cases to consider:
> 
> * What happens with numeric overflow scenarios?
> * Are negative numbers handled correctly in the approximation?
> * Need explicit test for approximation vs exact result validation.

I believe existing tests already validate approximate vs exact results. See `testApproximateVsExactQuery` in `ApproximatePointRangeQueryTests`. Edge cases like numeric overflows and negative numbers are handled appropriately in `ApproximatePointRangeQuery`.

> Nice integration with existing approximation framework. Consider:
> 
> * Should this be configurable per query? Some use cases may require exact numeric matching.
> * Memory overhead of maintaining both approximate and exact structures?

Approximation is more like a drop in replacement for regular queries since it doesn't impact correctness of queries. In my opinion, If we don't see regression in query performance I think it makes sense to make this a default.
There is a very minimal memory overhead for approximate queries since both the original query and approximate version are stored until execution but this is far outweighed by the performance benefit.

> Benchmark methodology is unclear - what dataset size/cardinality? Need to validate performance doesn't regress for small result sets where approximation overhead might dominate.

Yeah, that's my bad I wasn't super clear in my comment. I ran this benchmark on the http_logs dataset which is a relatively large dataset (~31g). That's a valid concern. `ApproximatePointRangeQuery` actually does contain a [small optimization](https://github.com/opensearch-project/OpenSearch/blob/4cd6962ff2f7932e6d59ab6213a4cc021e1112e8/server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java#L355) where it will fall back to `pointRangeQueryWeight.scorerSupplier(context)` if `size > values.size()`. Meaning for small datasets (less than 10k points), approximation is not utilized.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java [hunk 1]
```diff
@@ -1037,11 +1049,209 @@ private void testApproximateVsExactQuery(IndexSearcher searcher, String field, l
...
+                    Document doc = new Document();
+                    // Track the last value to use for DocValues
+                    long lastValue = 0;
+                    for (int j = 0; j < numValues; j++) {
+                        long randomValue = RandomNumbers.randomLongBetween(random(), 0, 200);
+                        numericType.addField(doc, numericType.fieldName, randomValue);
+                        lastValue = randomValue;
+                    }
+                    // Add DocValues field using the last value (required for sorting)
+                    numericType.addDocValuesField(doc, numericType.fieldName, lastValue);
```
Comment (atris) [inline server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java]: DocValues field is random and unrelated to the point values. If sorting is used anywhere in the test chain, this creates inconsistent behavior. Is this intended?

**Comment by sawansri** [comment]
> This is an interesting point. Is there an architectural limitation due to which this is not possible? If NumberFieldMapper.termQuery() returns ApproximatePointRangeQuery, why wouldn't it work inside a bool query? The query rewriting happens at the leaf level.
> 
> IMO this is a limitation that needs to be fixed to a wider real world applicability.

Yeah I agree with you, approximation within bool queries is definitely a big gap in our current framework. Here are the reasons we haven't done so already:

1. Conjunction Queries (FILTER and MUST) - All clauses must be constant score for us to even consider applying approximation (if we terminate at 10k hits, there is no guarantee that a later hit/conjunction combo scores higher). Additionally, even if all clauses are constant scoring, there is no guarantee that the conjunction of them will return 10k hits, hence why I proposed [this RFC](https://github.com/opensearch-project/OpenSearch/issues/18784).

2. Disjunction Queries (SHOULD and MUST_NOT) - Even if clauses are constant scoring, scores of clauses are summed essentially making the query scoring. We run into the same problem approximating here as conjunction queries that have scoring clauses.

> Approximation has correctness bounds - users are trading off accuracy for speed. It should be made clear and should be established by benchmarks.

So approximation in terms of our current implementation is actually a bit of a misnomer. We don't make any tradeoffs in accuracy for speedup. See [this issue](https://github.com/opensearch-project/OpenSearch/issues/13566) and [pr](https://github.com/opensearch-project/OpenSearch/pull/13788).

> http_logs is timestamp dependent. Consider benchmarking on datasets with patterns:
> 
> * Random integer distributions
> * Zipfian access patterns

Will do, thanks for the suggestions.

> * Actually measure the approximation error rates

As explained above, there is no error rate.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java (hunk 1 above)
Comment (sawansri) [reply server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java]: Hmm interesting, not sure why I was using DocValues there. Yeah, we should just be dealing with regular point values.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java [hunk 2]
```diff
@@ -988,9 +990,17 @@ private void testApproximateVsExactQuery(IndexSearcher searcher, String field, l
                 approxDocs.totalHits.value()
             );
         }
+
+        // If there are no matching documents, skip the sort tests
```
Comment (prudhvigodithi) [inline server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java]: Do we need to test this? if we index right range of docs do we ever end up in the scenario?

**Comment by prudhvigodithi** [comment]
@sawansri can you please check if have this query shape in our workloads https://github.com/opensearch-project/opensearch-benchmark-workloads/, we can run the benchmarks agains this PR.

File: modules/mapper-extras/src/test/java/org/opensearch/index/mapper/ScaledFloatFieldTypeTests.java [hunk 3]
```diff
@@ -68,7 +69,17 @@ public void testTermQuery() {
         long scaledValue = Math.round(value * ft.getScalingFactor());
         Query dvQuery = SortedNumericDocValuesField.newSlowExactQuery("scaled_float", scaledValue);
         Query query = new IndexOrDocValuesQuery(LongPoint.newExactQuery("scaled_float", scaledValue), dvQuery);
-        assertEquals(query, ft.termQuery(value, null));
+        Query expectedQuery = new ApproximateScoreQuery(
```
Comment (prudhvigodithi) [inline modules/mapper-extras/src/test/java/org/opensearch/index/mapper/ScaledFloatFieldTypeTests.java:72]: So we dont change anything in `ScaledFloatFieldMapper` because the `termQuery` uses the `NumberFieldMapper.NumberType.LONG.termQuery`

Comment (sawansri) [reply modules/mapper-extras/src/test/java/org/opensearch/index/mapper/ScaledFloatFieldTypeTests.java:72]: Yup, `ScaledFloatFieldMapper` implements a `termQuery` that uses `NumberFieldMapper.NumberType.LONG.termQuery()` which would fall under approximation.

**Comment by sawansri** [comment]
> @sawansri can you please check if have this query shape in our workloads [opensearch-project/opensearch-benchmark-workloads](https://github.com/opensearch-project/opensearch-benchmark-workloads/?rgh-link-date=2025-08-04T18%3A33%3A06Z), we can run the benchmarks agains this PR.

I don't see numeric term queries in any of the workloads. I'll probably have to run some benchmarks locally.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java (hunk 1 above)
Comment (prudhvigodithi) [reply server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java]: The tests in `ApproximatePointRangeQueryTests` index to both `BKD` and `doc_values`.
```
numericType.addField(doc, numericType.fieldName, i);
numericType.addDocValuesField(doc, numericType.fieldName, i);
```
This is because the `Sort sort = new Sort(new SortField(numericType.getSortFieldName(), numericType.getSortFieldType()));`, the `SortField` type does not directly support `half_float` and `unsigned_long` 

So to validate the results (same docs are returned) for `half_float` and `unsigned_long` we compare against the sort ran on `doc_values` (instead of BKD). If we run sorts on same field that is indexed in BKD we get the error as:

```
java.lang.IllegalArgumentException: Field half_float_field is indexed with 2 bytes per dimension, but org.apache.lucene.search.comparators.LongComparator@505e3f7e expected 8
```

**Comment by prudhvigodithi** [comment]
> I don't see numeric term queries in any of the workloads. I'll probably have to run some benchmarks locally.

Thanks and this is not a blocker for this PR, may be would consider adding it, here is the sample PR https://github.com/opensearch-project/opensearch-benchmark-workloads/pull/674 which I created in past to onboard new queries that are improved with approximation.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java (hunk 2 above)
Comment (sawansri) [reply server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java]: Yeah, I think that scenario never occurs in the tests.

**Comment by sawansri** [comment]
Ran some more benchmarks on Big5 (more balanced dataset):

Before Approximation
```
| Min Throughput | term-numeric | 1 | ops/s |
| Mean Throughput | term-numeric | 1.01 | ops/s |
| Median Throughput | term-numeric | 1.01 | ops/s |
| Max Throughput | term-numeric | 1.01 | ops/s |
| 50th percentile latency | term-numeric | 8.66894 | ms |
| 90th percentile latency | term-numeric | 9.52163 | ms |
| 99th percentile latency | term-numeric | 10.4394 | ms |
| 100th percentile latency | term-numeric | 11.7775 | ms |
| 50th percentile service time | term-numeric | 6.93358 | ms |
| 90th percentile service time | term-numeric | 7.49459 | ms |
| 99th percentile service time | term-numeric | 8.30141 | ms |
| 100th percentile service time | term-numeric | 9.75295 | ms |
| error rate | term-numeric | 0 | % |
```

After Approximation
```
| Min Throughput | term-numeric | 1 | ops/s |
| Mean Throughput | term-numeric | 1.01 | ops/s |
| Median Throughput | term-numeric | 1.01 | ops/s |
| Max Throughput | term-numeric | 1.01 | ops/s |
| 50th percentile latency | term-numeric | 8.53285 | ms |
| 90th percentile latency | term-numeric | 9.39101 | ms |
| 99th percentile latency | term-numeric | 10.4966 | ms |
| 100th percentile latency | term-numeric | 10.4971 | ms |
| 50th percentile service time | term-numeric | 6.77809 | ms |
| 90th percentile service time | term-numeric | 7.36977 | ms |
| 99th percentile service time | term-numeric | 8.40853 | ms |
| 100th percentile service time | term-numeric | 8.61667 | ms |
| error rate | term-numeric | 0 | % |
```

Query:
```
curl -X POST "http://localhost:9200/big5/_search" -H "Content-Type: application/json" -d '
{
    "query": 
        {
        "term": {
        "metrics.size": {
            "value": 1500
            }
        }
    }
}
' | jq '.'
```

Looks like we are seeing a slight latency drop. This makes sense since we expect to see the most benefit out of approximation on datasets where 10k hits can be reached early on in the BKD traversal (very common for range queries).

**Review by harshavamsi — APPROVED** [review]

**Comment by prudhvigodithi** [comment]
Coming from https://github.com/opensearch-project/OpenSearch/issues/18334 issue `Consider extending ApproximatePointRangeQuery to term queries. For low-cardinality field, this could yield better optimization results.` there was a request to extend for term queries, adding @kkewwei for some feedback

File: server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java [hunk 4]
```diff
@@ -392,7 +393,12 @@ public long cost() {
 
                             @Override
                             public Scorer get(long leadCost) throws IOException {
-                                intersectRight(values.getPointTree(), visitor, docCount);
+                                // Force term queries with desc sort through intersectLeft
+                                if (Arrays.equals(pointRangeQuery.getLowerPoint(), pointRangeQuery.getUpperPoint())) {
```
Comment (prudhvigodithi) [inline server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: How does it work with ? 
```
curl -X POST "http://localhost:9200/logs-*/_search" -H "Content-Type: application/json" -d '
{
  "query": {
    "term": {
      "status": {
        "value": 200
      }
    }
  },
  "sort": [
    {
      "@timestamp": {
        "order": "desc"
      }
    }
  ]
}
' | jq '.'
```

Comment (prudhvigodithi) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: Today we can run term with sort, I want to make sure the results does not break,

Comment (harshavamsi) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: This is not a good change

Comment (sawansri) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: > How does it work with ?
> 
> ```
> curl -X POST "http://localhost:9200/logs-*/_search" -H "Content-Type: application/json" -d '
> {
>   "query": {
>     "term": {
>       "status": {
>         "value": 200
>       }
>     }
>   },
>   "sort": [
>     {
>       "@timestamp": {
>         "order": "desc"
>       }
>     }
>   ]
> }
> ' | jq '.'
> ```

This change doesn't affect term queries where the primary sort field is different from the term query's field. In ApproximatePointRangeQuery's canApproximate, this statement gets executed. 
```
if (!primarySortField.fieldName().equals(pointRangeQuery.getField())) {
                    return false;
}
```

Comment (sawansri) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: Since there is a singular unique point value in a numeric term query and the tiebreaker ends up being [the docID](https://github.com/apache/lucene/blob/743eedec7b8010f16bdddac10bb8c0c4dd2cb231/lucene/core/src/java/org/apache/lucene/search/FieldValueHitQueue.java#L76), using intersectLeft results in the same top-k results as intersectRight. 

I'm also fine in disabling approximation for this specific case (term query with descending sort) due to performance concerns (queue has more comparisons to make) until intersectRight logic can be modified to return accurate hits for these kinds of queries.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java [hunk 5]
```diff
@@ -1046,7 +1054,7 @@ private void testApproximateVsExactQuery(IndexSearcher searcher, String field, N
             exactQuery = LongPoint.newRangeQuery(field, (long) lower, (long) upper);
         }
 
-        ApproximatePointRangeQuery approxQuery = new ApproximatePointRangeQuery(field, upperBytes, lowerBytes, dims, size, null, format);
+        ApproximatePointRangeQuery approxQuery = new ApproximatePointRangeQuery(field, lowerBytes, upperBytes, dims, size, null, format);
```
Comment (asimmahmood1) [inline server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java:1057]: Good find, but now I'm worried why didn't tests fail previously?

File: server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java (hunk 4 above)
Comment (asimmahmood1) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: If this is risky, can leave out this change and create an issue for it? I'm to merge in incremental changes, given there's enough test coverage.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java [hunk 6]
```diff
@@ -1146,6 +1154,172 @@ private void verifyRangeQueries(
...
+            }
+        }
+    }
+
+    public void testApproximateTermWithHighCardinality() throws IOException {
+        try (Directory directory = newDirectory()) {
+            try (RandomIndexWriter iw = new RandomIndexWriter(random(), directory, new WhitespaceAnalyzer())) {
+                int dims = 1;
+                int numDocs = 15000;
+                int targetValue = 1520;
```
Comment (asimmahmood1) [inline server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java:1209]: Should use random instead to get coverage for both postive and negative values? Or a selected set of targets, e.g. -neg, 0, pos, max, min? Maintainers, what the general guideline?

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java (hunk 5 above)
Comment (sawansri) [reply server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java:1057]: Hmmm, that's a good point. If the bounds are swapped, the query should be rewritten into a match no docs query. Would need to look into why the tests weren't failing previously.

File: server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java (hunk 4 above)
Comment (sawansri) [reply server/src/main/java/org/opensearch/search/approximate/ApproximatePointRangeQuery.java:397]: Yeah, that's what I was thinking as well. For the time being, we could disable approximation and send desc sorts on top level term queries through Lucene while this edge case is being sorted out.

File: server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java (hunk 6 above)
Comment (sawansri) [reply server/src/test/java/org/opensearch/search/approximate/ApproximatePointRangeQueryTests.java:1209]: Good idea, I can probably expand the test to change `numDocs` and `targetValue` to utilize random and include negative values, 0, max, min, etc.

**Comment by prudhvigodithi** [comment]
Just a follow up on this PR @sawansri, can you please check if you can limit the changes to ApproximatePointRangeQuery and have the top level term queries use the Approximation path ?
Thanks

**Comment by sawansri** [comment]
> Just a follow up on this PR @sawansri, can you please check if you can limit the changes to ApproximatePointRangeQuery and have the top level term queries use the Approximation path ? Thanks

Hey @prudhvigodithi , just to confirm, this would mean disabling approximation on top-level term queries with descending sort while letting term queries without sort and asc sort go through approximation. If this is the case, we would no longer need to force desc sort through `intersectLeft`.

**Comment by prudhvigodithi** [comment]
> Hey @prudhvigodithi , just to confirm, this would mean disabling approximation on top-level term queries with descending sort while letting term queries without sort and asc sort go through approximation.

I see, can the existing intersectRight work for the term queries when used desc sort ? Did I miss any thread?
